helper/resource/testing_test.go

```diff
@@ -1402,6 +1402,444 @@ func TestTestCheckResourceAttr(t *testing.T) {
+										"test_list_attribute.#": "1",
+										"test_list_attribute.0": "test-value",
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+			key:           "test_list_attribute.#",
+			value:         "2",
+			expectedError: fmt.Errorf("attribute 'test_list_attribute.#' expected '2', got '1'"),
+		},
+		"list attribute element count match 0 when empty": {
+			state: &terraform.State{
+				IsBinaryDrivenTest: true, // Always true now
+				Modules: []*terraform.ModuleState{
+					{
+						Path: []string{"root"},
+						Resources: map[string]*terraform.ResourceState{
+							"test_resource": {
+								Primary: &terraform.InstanceState{
+									Attributes: map[string]string{
+										"test_list_attribute.#": "0",
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+			key:   "test_list_attribute.#",
+			value: "0",
+		},
+		"list attribute element value match": {
+			state: &terraform.State{
+				IsBinaryDrivenTest: true, // Always true now
+				Modules: []*terraform.ModuleState{
+					{
+						Path: []string{"root"},
+						Resources: map[string]*terraform.ResourceState{
+							"test_resource": {
+								Primary: &terraform.InstanceState{
+									Attributes: map[string]string{
+										"test_list_attribute.#": "1",
+										"test_list_attribute.0": "test-value",
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+			key:   "test_list_attribute.0",
+			value: "test-value",
+		},
+		"list attribute element value mismatch": {
+			state: &terraform.State{
+				IsBinaryDrivenTest: true, // Always true now
+				Modules: []*terraform.ModuleState{
+					{
+						Path: []string{"root"},
+						Resources: map[string]*terraform.ResourceState{
+							"test_resource": {
+								Primary: &terraform.InstanceState{
+									Attributes: map[string]string{
+										"test_list_attribute.#": "1",
+										"test_list_attribute.0": "test-value",
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+			key:           "test_list_attribute.0",
+			value:         "not-test-value",
+			expectedError: fmt.Errorf("attribute 'test_list_attribute.0' expected 'not-test-value', got 'test-value'"),
+		},
+		"map attribute directly": {
+			state: &terraform.State{
+				IsBinaryDrivenTest: true, // Always true now
+				Modules: []*terraform.ModuleState{
+					{
+						Path: []string{"root"},
+						Resources: map[string]*terraform.ResourceState{
+							"test_resource": {
+								Primary: &terraform.InstanceState{
+									Attributes: map[string]string{
+										"test_map_attribute.%":        "1",
+										"test_map_attribute.testkey1": "test-value-1",
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+			key:           "test_map_attribute",
+			value:         "test-value",
+			expectedError: fmt.Errorf("map attribute 'test_map_attribute' must be checked by element count key (test_map_attribute.%%) or element value keys (e.g. test_map_attribute.examplekey)"),
```

Comment:
Should `test_map_attribute.%%` be `test_map_attribute.%`?

Reply:
In golang fmt strings `%%` it's the escape sequence for `%` I believe.

Reply:
🙃 more coffee required!